packages/dds/tree/src/feature-libraries/treeTextCursor.ts

```diff
@@ -167,34 +141,180 @@ export class RootedTextCursor extends TextCursor {
         // When navigating up, adjust cached anchor if present.
 
         let path: UpPath | undefined;
-        const length = this.indexStack.length;
-        assert(this.siblingStack.length === length, 0x34c /* Unexpected siblingStack.length */);
-        assert(this.keyStack.length === length + 1, 0x34d /* Unexpected keyStack.length */);
-        for (let height = 0; height < length; height++) {
+        // Skip top level, since root node in path is "undefined" and does not have a parent or index.
+        for (let height = 2; height < length; height += 2) {
+            const key = this.getStackedFieldKey(height - 1);
             path = {
                 parent: path,
-                parentIndex: this.indexStack[height],
-                parentField: this.keyStack[height],
+                parentIndex: this.getStackedNodeIndex(height),
+                parentField: key,
             };
         }
+
         path = {
             parent: path,
             parentIndex: this.index,
-            parentField: this.keyStack[length],
+            parentField: this.getStackedFieldKey(length - 1),
         };
         return path;
     }
+
+    public enterField(key: FieldKey): void {
+        // assert(this.mode === CursorLocationType.Nodes, "must be in nodes mode");
+        this.siblingStack.push(this.siblings);
+        this.indexStack.push(this.index);
+        this.index = -1;
+        this.siblings = key;
+    }
+
+    public get mode(): CursorLocationType {
+        return this.siblingStack.length % 2 === 0 ? CursorLocationType.Nodes : CursorLocationType.Fields;
+    }
+
+    public get pending(): boolean {
+        return false;
```

Comment:
This cursor never is pending, so that can't happen.

For cursors in general, it is invalid to navigate into a pending node or field. The API specifies this but currently leaves this behavior in this invalid case undefined (all the APIs which could do such a navigation are commented indicating they are invalid to call when pending), though we might choose to throw a specific exception for it (at least in some cursor sub-interfaces) in the future instead of requiring everyone to check for it.

Reply:
Did you mean it is invalid to descend **from** a pending node or field?  (I figured the way pending becomes 'true' is that you have navigated into a pending field/node.)

Reply:
@CraigMacomber